<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Advanced Phenotyping in Single Cell Data Analysis • scdneyAdvancedPhenotyping</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Advanced Phenotyping in Single Cell Data Analysis">
<meta property="og:description" content="scdneyAdvancedPhenotyping">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scdneyAdvancedPhenotyping</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/advanced_phenotyping.html">Advanced Phenotyping in Single Cell Data Analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/SydneyBioX/scdneyAdvancedPhenotyping">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="advanced_phenotyping_files/header-attrs-2.10/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Advanced Phenotyping in Single Cell Data Analysis</h1>
                        <h4 class="author">Yingxin Lin<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> Hani Jieun Kim<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> Carissa Chen<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/scdneyAdvancedPhenotyping/blob/master/vignettes/advanced_phenotyping.Rmd"><code>vignettes/advanced_phenotyping.Rmd</code></a></small>
      <div class="hidden name"><code>advanced_phenotyping.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<div id="description" class="section level3">
<h3 class="hasAnchor">
<a href="#description" class="anchor"></a>Description</h3>
<p>[Brief description of workshop]</p>
</div>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<p>List any workshop prerequisites, for example:</p>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Familiarity with the GenomicRanges class</li>
<li>Familiarity with xyz vignette (provide link)</li>
</ul>
<p>List relevant background reading for the workshop, including any theoretical background you expect students to have.</p>
<ul>
<li>List any textbooks, papers, or other reading that students should be familiar with. Include direct links where possible.</li>
</ul>
</div>
<div id="participation" class="section level3">
<h3 class="hasAnchor">
<a href="#participation" class="anchor"></a>Participation</h3>
<p>Describe how students will be expected to participate in the workshop.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<p>List any <em>R</em> / <em>Bioconductor</em> packages that will be explicitly covered.</p>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<p>Structure of the 2-hour workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Packages</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Package Development</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Contributing to Bioconductor</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Best Practices</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
<div id="workshop-goals-and-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h3>
<p>List “big picture” student-centered workshop goals and learning objectives. Learning goals and objectives are related, but not the same thing. These goals and objectives will help some people to decide whether to attend the conference for training purposes, so please make these as precise and accurate as possible.</p>
<p><em>Learning goals</em> are high-level descriptions of what participants will learn and be able to do after the workshop is over. <em>Learning objectives</em>, on the other hand, describe in very specific and measurable terms specific skills or knowledge attained. The <a href="#bloom">Bloom’s Taxonomy</a> may be a useful framework for defining and describing your goals and objectives, although there are others.</p>
</div>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h3>
<p>Some examples:</p>
<ul>
<li>describe how to…</li>
<li>identify methods for…</li>
<li>understand the difference between…</li>
</ul>
</div>
<div id="learning-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h3>
<ul>
<li>analyze xyz data to produce…</li>
<li>create xyz plots</li>
<li>evaluate xyz data for artifacts</li>
</ul>
</div>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>1. Introduction</h2>
<p>General introduction to the advanced phenotyping course (e.g., summary of the introduction PPT, Carissa to fill?).</p>
</div>
<div id="defining-cellular-identity" class="section level2">
<h2 class="hasAnchor">
<a href="#defining-cellular-identity" class="anchor"></a>2. Defining cellular identity</h2>
<div id="load-libraries" class="section level3">
<h3 class="hasAnchor">
<a href="#load-libraries" class="anchor"></a>2.1 Load libraries</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Cepo</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">slingshot</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jrnold/ggthemes">ggthemes</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/ggridges/">ggridges</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape">reshape2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/">ggpubr</a></span><span class="op">)</span>

<span class="fu">theme_set</span><span class="op">(</span><span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="loading-preprocessed-data" class="section level3">
<h3 class="hasAnchor">
<a href="#loading-preprocessed-data" class="anchor"></a>2.2 Loading preprocessed data</h3>
<p>About the data… from <a href="https://www.nature.com/articles/s41586-018-0414-6#additional-information">this paper</a>. We provide the preprocessed version in this workshop. The original data can be access from GEO with the accession number XXXXXX.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># code to load data </span>
<span class="co"># we can also run some of the processing steps here (e.g., log transform, filter lowly expressed genes etc.)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/sce_DentateGyrus_hippocampus.rds"</span><span class="op">)</span>
<span class="va">sce</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 27998 18213 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(3): counts spliced unspliced</span>
<span class="co">#&gt; rownames(27998): Xkr4 Gm37381 ... Gm20837 Erdr1</span>
<span class="co">#&gt; rowData names(6): Accession Chromosome ... Start Strand</span>
<span class="co">#&gt; colnames(18213): 10X83_2:AAACGGGGTCTCTTTAx 10X83_2:AACCATGGTTCAACCAx</span>
<span class="co">#&gt;   ... 10X84_3:TTTGTCAAGCGTCAAGx 10X84_3:TTTCCTCGTGAAAGAGx</span>
<span class="co">#&gt; colData names(32): Age CellID ... cDNA_Lib_Ok ngperul_cDNA</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; mainExpName: NULL</span>
<span class="co">#&gt; altExpNames(0):</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"Nbl1|Nbl2|Granule|CA"</span>, <span class="va">sce</span><span class="op">$</span><span class="va">ClusterName</span><span class="op">)</span><span class="op">]</span>
<span class="va">sce</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 27998 12935 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(3): counts spliced unspliced</span>
<span class="co">#&gt; rownames(27998): Xkr4 Gm37381 ... Gm20837 Erdr1</span>
<span class="co">#&gt; rowData names(6): Accession Chromosome ... Start Strand</span>
<span class="co">#&gt; colnames(12935): 10X83_2:AAACGGGGTCTCTTTAx 10X83_2:AACCATGGTTCAACCAx</span>
<span class="co">#&gt;   ... 10X84_3:TTTGTCAAGCGTCAAGx 10X84_3:TTTCCTCGTGAAAGAGx</span>
<span class="co">#&gt; colData names(32): Age CellID ... cDNA_Lib_Ok ngperul_cDNA</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; mainExpName: NULL</span>
<span class="co">#&gt; altExpNames(0):</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
</div>
<div id="visualising-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#visualising-the-data" class="anchor"></a>2.3 Visualising the data</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># code to run PCA/UMAP/TSNE</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runDiffusionMap.html">runDiffusionMap</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>


<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"ClusterName"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"27998 genes"</span><span class="op">)</span> 
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"ClusterName"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"27998 genes"</span><span class="op">)</span>
<span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"ClusterName"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"27998 genes"</span><span class="op">)</span>
<span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotDiffusionMap</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"ClusterName"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"643 Cepo genes"</span><span class="op">)</span>


<span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, <span class="va">p4</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">2</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-4-1.png" width="700"> [give some example marker genes for each cell type]</p>
</div>
<div id="clustering-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#clustering-the-data" class="anchor"></a>2.3 Clustering the data</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># code to perform clustering of the data</span></code></pre></div>
</div>
<div id="identifying-the-cell-types" class="section level3">
<h3 class="hasAnchor">
<a href="#identifying-the-cell-types" class="anchor"></a>2.3 Identifying the cell types</h3>
<p>Now, we want to identify a cluster to a cell type. This task is easy to apply when discrete and well-annotated cell types, i.e., those with well-known markers, are present in your data. The procedure becomes more challenging when we are met with a data with rare or unknown cell types or transitional cell states. Either way, the general idea is to find a set of genes that are expressed different in a cell compared to the rest of the cells in the dataset. Thus, gene sets must be identifying of the particular cell type of the cell.</p>
<p>There are many methods</p>
<div id="cepo-to-identify-cell-identity-genes" class="section level4">
<h4 class="hasAnchor">
<a href="#cepo-to-identify-cell-identity-genes" class="anchor"></a>Cepo to identify cell identity genes</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Code to run Cepo</span>
<span class="va">cepo</span> <span class="op">=</span> <span class="fu">Cepo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Cepo/man/Cepo.html">Cepo</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">sce</span><span class="op">$</span><span class="va">ClusterName</span>, exprsPct <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="va">cepoGenes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu">Cepo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Cepo/man/topGenes.html">topGenes</a></span><span class="op">(</span><span class="va">cepo</span>, n <span class="op">=</span> <span class="fl">100</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">sce</span> <span class="op">=</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, subset_row <span class="op">=</span> <span class="va">cepoGenes</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runDiffusionMap.html">runDiffusionMap</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>

<span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"ClusterName"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"643 Cepo genes"</span><span class="op">)</span> 
<span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"ClusterName"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"643 Cepo genes"</span><span class="op">)</span>
<span class="va">p3</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"ClusterName"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"643 Cepo genes"</span><span class="op">)</span>
<span class="va">p4</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotDiffusionMap</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"ClusterName"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"643 Cepo genes"</span><span class="op">)</span>

<span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, <span class="va">p4</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">2</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Manual annotation using markers (and check original label)</p>
</div>
<div id="examine-expression-of-cepo-genes-in-the-low-dimensional-space" class="section level4">
<h4 class="hasAnchor">
<a href="#examine-expression-of-cepo-genes-in-the-low-dimensional-space" class="anchor"></a>Examine expression of Cepo genes in the low dimensional space</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Code to visualise markers</span></code></pre></div>
</div>
<div id="violin-plots" class="section level4">
<h4 class="hasAnchor">
<a href="#violin-plots" class="anchor"></a>Violin plots</h4>
<p>Violin plots enable us to examine the distribution of the expression values associated with each cell in the clusters. The x-axis denote the clusters, and the y-axis represents the log-transformed gene expression of a given gene. Alternatives of violin plots are jitter and swarmplots, which unlike violin plots show each cell as a point.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Code to plot violin plot</span></code></pre></div>
</div>
<div id="saving-outputs" class="section level4">
<h4 class="hasAnchor">
<a href="#saving-outputs" class="anchor"></a>Saving outputs</h4>
<p>Finally, we can save our marker gene analysis for use later. These top marker genes can be used to subset the data prior to performing classification of the data</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Code to save top X genes </span></code></pre></div>
<p>How we can use these top genes for classification.</p>
</div>
</div>
</div>
<div id="trajectory-inference-how-can-i-infer-the-trajectory-for-my-data" class="section level2">
<h2 class="hasAnchor">
<a href="#trajectory-inference-how-can-i-infer-the-trajectory-for-my-data" class="anchor"></a>3. Trajectory inference: How can I infer the trajectory for my data?</h2>
<div id="how-to-build-a-trajectory" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-build-a-trajectory" class="anchor"></a>How to build a trajectory?</h3>
<p>Construct a trajectory using <code>slingshot</code> on UMAP reduced dimension space.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">slingshot_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingshot.html">slingshot</a></span><span class="op">(</span><span class="va">sce</span>, 
                            clusterLabels <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">ClusterName</span>,
                            reducedDim <span class="op">=</span> <span class="st">'UMAP'</span>,
                            start.clus <span class="op">=</span> <span class="st">"Nbl1"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## visualising using base R</span>
<span class="va">rd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">slingshot_umap</span>, <span class="st">"UMAP"</span><span class="op">)</span>
<span class="va">plotcol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/tableau_color_pal.html">tableau_color_pal</a></span><span class="op">(</span><span class="st">"Classic 10"</span><span class="op">)</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">slingshot_umap</span><span class="op">$</span><span class="va">ClusterName</span><span class="op">)</span><span class="op">]</span>

<span class="co"># set the aspect ratio for base R plot as 1</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>pty <span class="op">=</span> <span class="st">"s"</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">rd</span>, col <span class="op">=</span> <span class="va">plotcol</span>, 
     pch <span class="op">=</span> <span class="fl">16</span>, asp <span class="op">=</span> <span class="fl">1</span>, xlab <span class="op">=</span> <span class="st">"UMAP1"</span>, ylab <span class="op">=</span> <span class="st">"UMAP2"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/SlingshotDataSet.html">SlingshotDataSet</a></span><span class="op">(</span><span class="va">slingshot_umap</span><span class="op">)</span>, lwd<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="st">'black'</span><span class="op">)</span></code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## visualising using ggplot </span>
<span class="va">df_toPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>UMAP1 <span class="op">=</span> <span class="va">rd</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,
                        UMAP2 <span class="op">=</span> <span class="va">rd</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,
                        ClusterName <span class="op">=</span> <span class="va">slingshot_umap</span><span class="op">$</span><span class="va">ClusterName</span><span class="op">)</span>
<span class="va">pseudotime_umap</span> <span class="op">&lt;-</span> <span class="fu">slingshot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingPseudotime.html">slingPseudotime</a></span><span class="op">(</span><span class="va">slingshot_umap</span><span class="op">)</span>
<span class="va">df_toPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">df_toPlot</span>, <span class="va">pseudotime_umap</span><span class="op">)</span>

<span class="va">curves</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingCurves.html">slingCurves</a></span><span class="op">(</span><span class="va">slingshot_umap</span>, as.df <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span>data <span class="op">=</span> <span class="va">df_toPlot</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">UMAP1</span>, y <span class="op">=</span> <span class="va">UMAP2</span>, color <span class="op">=</span> <span class="va">ClusterName</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_path</span><span class="op">(</span>data <span class="op">=</span> <span class="va">curves</span>,
            <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, group <span class="op">=</span> <span class="va">Lineage</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html">scale_color_tableau</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Classic 10"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## visualising the psuedotime</span>
<span class="va">df_toPlot2</span> <span class="op">&lt;-</span> <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span><span class="op">(</span><span class="va">df_toPlot</span>, id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"UMAP1"</span>, <span class="st">"UMAP2"</span>, <span class="st">"ClusterName"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">df_toPlot2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"UMAP1"</span>, <span class="st">"UMAP2"</span>, <span class="st">"ClusterName"</span>, <span class="st">"Lineages"</span>, <span class="st">"Pseudotime"</span><span class="op">)</span>
<span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span>data <span class="op">=</span> <span class="va">df_toPlot2</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">UMAP1</span>, y <span class="op">=</span> <span class="va">UMAP2</span>, color <span class="op">=</span> <span class="va">Pseudotime</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_path</span><span class="op">(</span>data <span class="op">=</span> <span class="va">curves</span>,
            <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, group <span class="op">=</span> <span class="va">Lineage</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_color_viridis_c</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">Lineages</span><span class="op">)</span></code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://wilkelab.org/ggridges/reference/geom_density_ridges.html">geom_density_ridges2</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">df_toPlot2</span>, 
                       <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Pseudotime</span>, y <span class="op">=</span> <span class="va">ClusterName</span>, fill <span class="op">=</span> <span class="va">ClusterName</span><span class="op">)</span>, 
                       alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html">scale_fill_tableau</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Classic 10"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">Lineages</span><span class="op">)</span> </code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p><strong>Quick quiz:</strong></p>
<ul>
<li>Do these orderings make sense given the cluster labels? Why?</li>
<li>Would specifying starting cluster (<code>start.clus</code>) or not impact our trajectory results?</li>
</ul>
</div>
<div id="how-does-the-choice-of-dimensionality-reduction-affect-the-trajectory-inference" class="section level3">
<h3 class="hasAnchor">
<a href="#how-does-the-choice-of-dimensionality-reduction-affect-the-trajectory-inference" class="anchor"></a>How does the choice of dimensionality reduction affect the trajectory inference?</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Perform Slingshot analysis using TSNE</span>

<span class="va">slingshot_tsne</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingshot.html">slingshot</a></span><span class="op">(</span><span class="va">sce</span>, 
                            clusterLabels <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">ClusterName</span>,
                            reducedDim <span class="op">=</span> <span class="st">"TSNE"</span>,
                            start.clus <span class="op">=</span> <span class="st">"Nbl1"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rd_tsne</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">slingshot_tsne</span>, <span class="st">"TSNE"</span><span class="op">)</span>
<span class="va">df_toPlot</span><span class="op">$</span><span class="va">TSNE1</span> <span class="op">&lt;-</span> <span class="va">rd_tsne</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>
<span class="va">df_toPlot</span><span class="op">$</span><span class="va">TSNE2</span> <span class="op">&lt;-</span> <span class="va">rd_tsne</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>
<span class="va">curves_tsne</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingCurves.html">slingCurves</a></span><span class="op">(</span><span class="va">slingshot_tsne</span>, as.df <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">pseudotime_tsne</span> <span class="op">&lt;-</span> <span class="fu">slingshot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingPseudotime.html">slingPseudotime</a></span><span class="op">(</span><span class="va">slingshot_tsne</span><span class="op">)</span>

<span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span>data <span class="op">=</span> <span class="va">df_toPlot</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">TSNE1</span>, y <span class="op">=</span> <span class="va">TSNE2</span>, color <span class="op">=</span> <span class="va">ClusterName</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_path</span><span class="op">(</span>data <span class="op">=</span> <span class="va">curves_tsne</span>,
            <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Dim.1</span>, y <span class="op">=</span> <span class="va">Dim.2</span>, group <span class="op">=</span> <span class="va">Lineage</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html">scale_color_tableau</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Classic 10"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>pty <span class="op">=</span> <span class="st">"s"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">pseudotime_tsne</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="va">pseudotime_umap</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">pseudotime_tsne</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="va">pseudotime_umap</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, use <span class="op">=</span> <span class="st">"complete.obs"</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.9968337</span></code></pre></div>
<p>We can see that the trajectories the built based on UMAP and tSNE reduced dimension space are highly correlated with each other (cor &gt; 0.99).</p>
<p><strong>Quick quiz:</strong></p>
<ul>
<li>Would dimension reduction using different number of selected features impact the trajecotry results?</li>
</ul>
</div>
<div id="how-well-do-the-pseudotime-orderings-correlate-from-different-trajectory-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#how-well-do-the-pseudotime-orderings-correlate-from-different-trajectory-methods" class="anchor"></a>How well do the pseudotime orderings correlate from different trajectory methods?</h3>
<p>Here we will use <code>monocle3</code> to construct the trajectory on the same UMAP space.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(monocle3)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Construt cell data set object for monocle3</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce) <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(<span class="fu">rowData</span>(sce), <span class="at">gene_short_name =</span> <span class="fu">rownames</span>(sce))</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>cds <span class="ot">&lt;-</span> <span class="fu">as</span>(sce, <span class="st">"cell_data_set"</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>cds <span class="ot">&lt;-</span> <span class="fu">cluster_cells</span>(cds, <span class="at">reduction_method =</span> <span class="st">"UMAP"</span>)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>cds <span class="ot">&lt;-</span> <span class="fu">learn_graph</span>(cds, <span class="at">use_partition =</span> <span class="cn">FALSE</span>) <span class="co"># set use_partition to FALSE so only a single graph is learnt</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_cells</span>(cds,</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>           <span class="at">color_cells_by =</span> <span class="st">"ClusterName"</span>,</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>           <span class="at">label_groups_by_cluster =</span> <span class="cn">FALSE</span>,</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>           <span class="at">label_leaves =</span> <span class="cn">FALSE</span>,</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>           <span class="at">label_branch_points =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_tableau</span>(<span class="at">palette =</span> <span class="st">"Classic 10"</span>) <span class="sc">+</span></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">aspect.ratio =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R">

<span class="va">get_earliest_principal_node</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">cds</span>, <span class="va">coldata_name</span>, <span class="va">start.clust</span><span class="op">)</span><span class="op">{</span>
  <span class="va">cell_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">[</span>, <span class="va">coldata_name</span><span class="op">]</span> <span class="op">==</span> <span class="va">start.clust</span><span class="op">)</span>
  <span class="va">closest_vertex</span> <span class="op">&lt;-</span> <span class="va">cds</span><span class="op">@</span><span class="va">principal_graph_aux</span><span class="op">[[</span><span class="st">"UMAP"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">pr_graph_cell_proj_closest_vertex</span>
  <span class="va">closest_vertex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">closest_vertex</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span>
  <span class="va">root_pr_nodes</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/V.html">V</a></span><span class="op">(</span><span class="fu">principal_graph</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">[[</span><span class="st">"UMAP"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">closest_vertex</span><span class="op">[</span><span class="va">cell_ids</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
  <span class="va">root_pr_nodes</span>
<span class="op">}</span>

<span class="co"># Set the root cells</span>
<span class="va">root_cells</span> <span class="op">&lt;-</span> <span class="fu">get_earliest_principal_node</span><span class="op">(</span><span class="va">cds</span>, coldata_name <span class="op">=</span> <span class="st">"ClusterName"</span>,
                                          start.clust <span class="op">=</span> <span class="st">"Nbl1"</span><span class="op">)</span>
<span class="co"># Order the cells to calculate pseudotime</span>
<span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">order_cells</span><span class="op">(</span><span class="va">cds</span>, 
                   root_pr_nodes <span class="op">=</span> <span class="va">root_cells</span><span class="op">)</span>


<span class="fu">plot_cells</span><span class="op">(</span><span class="va">cds</span>,
           color_cells_by <span class="op">=</span> <span class="st">"pseudotime"</span>,
           label_cell_groups <span class="op">=</span> <span class="cn">FALSE</span>,
           label_leaves <span class="op">=</span> <span class="cn">FALSE</span>,
           label_branch_points <span class="op">=</span> <span class="cn">FALSE</span>,
           graph_label_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-18-2.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">monocle3_pseudotime</span> <span class="op">&lt;-</span> <span class="va">cds</span><span class="op">@</span><span class="va">principal_graph_aux</span><span class="op">[[</span><span class="st">"UMAP"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">pseudotime</span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>pty <span class="op">=</span> <span class="st">"s"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">monocle3_pseudotime</span>, <span class="va">pseudotime_umap</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, xlab <span class="op">=</span> <span class="st">"Monocle 3"</span>, ylab <span class="op">=</span> <span class="st">"Slingshot (Lineage 1)"</span><span class="op">)</span></code></pre></div>
<p><img src="advanced_phenotyping_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">monocle3_pseudotime</span>, <span class="va">pseudotime_umap</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, use <span class="op">=</span> <span class="st">"complete.obs"</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.9969634</span></code></pre></div>
</div>
<div id="extension-rna-velocity-vs-trajectory-pseudotime-inference" class="section level3">
<h3 class="hasAnchor">
<a href="#extension-rna-velocity-vs-trajectory-pseudotime-inference" class="anchor"></a>Extension: RNA velocity vs trajectory pseudotime inference</h3>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">velocyto.R</span><span class="op">)</span>
<span class="va">emat</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"spliced"</span><span class="op">)</span>
<span class="va">nmat</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"unspliced"</span><span class="op">)</span>
<span class="va">cluster.label</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">ClusterName</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cluster.label</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="co"># Use UMAP embeddings</span>
<span class="va">emb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"UMAP"</span><span class="op">)</span>
<span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span>
<span class="co"># Calculate distance</span>
<span class="va">cell.dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/pkg/velocyto.R/man/armaCor.html">armaCor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Filtering the genes</span>
<span class="va">emat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/velocyto.R/man/filter.genes.by.cluster.expression.html">filter.genes.by.cluster.expression</a></span><span class="op">(</span><span class="va">emat</span>, <span class="va">cluster.label</span>, min.max.cluster.average <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>
<span class="va">nmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/velocyto.R/man/filter.genes.by.cluster.expression.html">filter.genes.by.cluster.expression</a></span><span class="op">(</span><span class="va">nmat</span>, <span class="va">cluster.label</span>, min.max.cluster.average <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>

<span class="co"># Using 3838 genes for velocity estimation</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">emat</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">emat</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">rvel.cd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/velocyto.R/man/gene.relative.velocity.estimates.html">gene.relative.velocity.estimates</a></span><span class="op">(</span><span class="va">emat</span>, <span class="va">nmat</span>, deltaT <span class="op">=</span> <span class="fl">1</span>,
                                            kCells <span class="op">=</span> <span class="fl">20</span>, cell.dist <span class="op">=</span> <span class="va">cell.dist</span>,
                                            fit.quantile <span class="op">=</span> <span class="fl">0.02</span>, n.cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/velocyto.R/man/show.velocity.on.embedding.cor.html">show.velocity.on.embedding.cor</a></span><span class="op">(</span><span class="va">emb</span>, <span class="va">rvel.cd</span>, n <span class="op">=</span> <span class="fl">300</span>, scale <span class="op">=</span> <span class="st">'sqrt'</span>,
                               <span class="co"># cell.colors = ac(cell.colors, alpha=0.5), </span>
                               cex <span class="op">=</span> <span class="fl">0.8</span>, arrow.scale <span class="op">=</span> <span class="fl">5</span>, show.grid.flow <span class="op">=</span> <span class="cn">TRUE</span>,
                               min.grid.cell.mass <span class="op">=</span> <span class="fl">0.5</span>, grid.n <span class="op">=</span> <span class="fl">40</span>,
                               arrow.lwd <span class="op">=</span> <span class="fl">1</span>, do.par <span class="op">=</span> <span class="cn">F</span>, cell.border.alpha <span class="op">=</span> <span class="fl">0.1</span>,
                               n.cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="how-can-i-characterise-the-trajectory-of-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#how-can-i-characterise-the-trajectory-of-the-data" class="anchor"></a>4. How can I characterise the trajectory of the data?</h2>
<div id="what-are-the-marker-genes-associated-with-the-different-lineagespseudotime" class="section level3">
<h3 class="hasAnchor">
<a href="#what-are-the-marker-genes-associated-with-the-different-lineagespseudotime" class="anchor"></a>What are the marker genes associated with the different lineages/pseudotime?</h3>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## TradeSeq</span></code></pre></div>
</div>
<div id="what-are-the-different-patterns-of-gene-expression-along-pseudotime" class="section level3">
<h3 class="hasAnchor">
<a href="#what-are-the-different-patterns-of-gene-expression-along-pseudotime" class="anchor"></a>What are the different patterns of gene expression along pseudotime?</h3>
</div>
<div id="are-these-patterns-associated-with-the-pathways-associated-with-the-trajectory-de" class="section level3">
<h3 class="hasAnchor">
<a href="#are-these-patterns-associated-with-the-pathways-associated-with-the-trajectory-de" class="anchor"></a>Are these patterns associated with the pathways associated with the trajectory DE?</h3>
</div>
<div id="what-does-gene-pair-correlation-change-across-lineages" class="section level3">
<h3 class="hasAnchor">
<a href="#what-does-gene-pair-correlation-change-across-lineages" class="anchor"></a>What does gene pair correlation change across lineages?</h3>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># scHOT</span></code></pre></div>
</div>
</div>
<div id="cell-cell-communication" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-cell-communication" class="anchor"></a>5. Cell-cell communication</h2>
</div>
<div id="conclusions-and-wrap-up" class="section level2">
<h2 class="hasAnchor">
<a href="#conclusions-and-wrap-up" class="anchor"></a>6. Conclusions and wrap-up</h2>
</div>
<div id="reference-and-other-useful-resources" class="section level2">
<h2 class="hasAnchor">
<a href="#reference-and-other-useful-resources" class="anchor"></a>Reference and other useful resources:</h2>
<ul>
<li>dynverse, a collection of R packages aimed at supporting the trajectory inference: <a href="https://dynverse.org" class="uri">https://dynverse.org</a>
</li>
<li>RNA velocity: <a href="http://velocyto.org" class="uri">http://velocyto.org</a>
</li>
</ul>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:yinxin.lin@sydney.edu.au" class="email">yinxin.lin@sydney.edu.au</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p><a href="mailto:hani.kim@sydney.edu.au" class="email">hani.kim@sydney.edu.au</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p><a href="mailto:cchen@uni.sydney.edu.au" class="email">cchen@uni.sydney.edu.au</a><a href="#fnref3" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Sydney Precision Bioinformatics Alliance.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
